### ISSUE TITLE
Achieving arcane affinity 50%+ depth permakicks affected player from a dedicated server

### ISSUE THREAD

**Issue by Not-February** [AirBurn0/am2playground#23]
First off, love that you've made this mod; thus far it's added a lot of entertaining replay value to a mod I've enjoyed for years. So thanks for that! Unfortunately, I've run into a bit of a snag and thought I'd check if it's something that might need fixing.

I've been playing with this mod on a dedicated server and as soon as I reached 50% arcane affinity from casting a spell, I got booted and started getting an 'Internal Server Error' message every time I tried to reconnect. It took me a bit, but on a hunch after reading the error in the server log I edited my player NBT data to set my arcane affinity to a value below 50% and sure enough, that allowed me to reconnect again without issue.

This effectively means that in multiplayer, you'll lock yourself out of your character if you get arcane affinity to 50%+. Which is... not ideal.

I can't seem to find any publicly available information that mentions AM2PG actually changing the effects of the arcane affinity in any way, so this came as something of a surprise. 

I also can't reproduce the problem in single player, but the fact that the error mentions client-specific code in the server log probably implicates it as a multiplayer-specific issue.

It is worth noting for the sake of being thorough that I am currently using [this fork](https://github.com/RubilaxXxx/ArsMagica2-5/releases) of AM2.5 (using v1.6.9-RE)for the config option for disabling burnout(as a means if avoiding [this issue](https://github.com/TCLProject/ArsMagica2-5/issues/52).

I know it is technically an unsupported version of AM2.5 for AM2PG, but I'm hoping if nothing else that you can at least tell me if this is a compatibility issue between the fork of AM2.5 and AM2PG or if it is in fact an issue with AM2PG that could happen regardless of what version of AM2.5 is used. I can set aside a time when no one is playing to specifically test a supported version of AM2.5 if you believe it to be needed.

AM2PG mod version: 0.11.3

I can share the full server log and mod list as well if needed, but this the most relevant part that occurs when I was trying to reconnect to the server(on the client, it kicked me with the message **Internal Server Error**):

```
[2024-09-30 00:34:47] [Server thread/WARN]: Failed to handle packet for (IP removed for privacy)
[2024-09-30 00:34:47]: net.minecraft.util.ReportedException: Ticking player
[2024-09-30 00:34:47]: at net.minecraft.entity.player.EntityPlayerMP.func_71127_g(EntityPlayerMP.java:349) ~[mw.class:?]
[2024-09-30 00:34:47]: at net.minecraft.network.NetHandlerPlayServer.func_147347_a(NetHandlerPlayServer.java:303) ~[nh.class:?]
[2024-09-30 00:34:47]: at net.minecraft.network.play.client.C03PacketPlayer.func_148833_a(SourceFile:137) ~[jd.class:?]
[2024-09-30 00:34:47]: at net.minecraft.network.play.client.C03PacketPlayer$C06PacketPlayerPosLook.func_148833_a(SourceFile:20) ~[jf.class:?]
[2024-09-30 00:34:47]: at net.minecraft.network.NetworkManager.func_74428_b(NetworkManager.java:212) ~[ej.class:?]
[2024-09-30 00:34:47]: at net.minecraft.network.NetworkSystem.func_151269_c(NetworkSystem.java:165) [nc.class:?]
[2024-09-30 00:34:47]: at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:659) [MinecraftServer.class:?]
[2024-09-30 00:34:47]: at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:334) [lt.class:?]
[2024-09-30 00:34:47]: at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:547) [MinecraftServer.class:?]
[2024-09-30 00:34:47]: at [net.minecraft.server.MinecraftServer.run](http://net.minecraft.server.minecraftserver.run/)(MinecraftServer.java:427) [MinecraftServer.class:?]
[2024-09-30 00:34:47]: at [net.minecraft.server.MinecraftServer$2.run](http://net.minecraft.server.minecraftserver$2.run/)(MinecraftServer.java:685) [li.class:?]
[2024-09-30 00:34:47]: Caused by: java.lang.NoClassDefFoundError: net/minecraft/client/Minecraft
[2024-09-30 00:34:47]: at airburn.am2playground.utils.ArcaneEffectHandler.getRandomPotion(ArcaneEffectHandler.java:33) ~[ArcaneEffectHandler.class:?]
[2024-09-30 00:34:47]: at airburn.am2playground.event.ForgeEventHandler.arcaneEffects(ForgeEventHandler.java:539) ~[ForgeEventHandler.class:?]
[2024-09-30 00:34:47]: at cpw.mods.fml.common.eventhandler.ASMEventHandler_920_ForgeEventHandler_arcaneEffects_LivingUpdateEvent.invoke(.dynamic) ~[?:?]
[2024-09-30 00:34:47]: at cpw.mods.fml.common.eventhandler.ASMEventHandler.invoke(ASMEventHandler.java:54) ~[ASMEventHandler.class:?]
[2024-09-30 00:34:47]: at [cpw.mods.fml.common.eventhandler.EventBus.post](http://cpw.mods.fml.common.eventhandler.eventbus.post/)(EventBus.java:140) ~[EventBus.class:?]
[2024-09-30 00:34:47]: at net.minecraftforge.common.ForgeHooks.onLivingUpdate(ForgeHooks.java:287) ~[ForgeHooks.class:?]
[2024-09-30 00:34:47]: at net.minecraft.entity.EntityLivingBase.func_70071_h_(EntityLivingBase.java:1560) ~[sv.class:?]
[2024-09-30 00:34:47]: at net.minecraft.entity.player.EntityPlayer.func_70071_h_(EntityPlayer.java:288) ~[yz.class:?]
[2024-09-30 00:34:47]: at net.minecraft.entity.player.EntityPlayerMP.func_71127_g(EntityPlayerMP.java:295) ~[mw.class:?]```

**Comment by AirBurn0** [comment]
# TL:DR
```java
 ((WeightedRandomPotion)WeightedRandom.getRandomItem(Minecraft.getMinecraft().theWorld.rand, arcaneEffects)).getPotion()
```
DAMN I F*CKED UP AGAIN 😭
(`Minecraft.getMinecraft().theWorld` is only client-side method; It will crash on server.)
Thanks for finding it, the fix will be made in a minute and released on GitHub... well, when I check that the bug is gone.

# About what you said
> I've been playing with this mod on a dedicated server and as soon as I reached 50% arcane affinity from casting a spell, I got booted and started getting an 'Internal Server Error' message every time I tried to reconnect. It took me a bit, but on a hunch after reading the error in the server log I edited my player NBT data to set my arcane affinity to a value below 50% and sure enough, that allowed me to reconnect again without issue.
> 
> I can't seem to find any publicly available information that mentions AM2PG actually changing the effects of the arcane affinity in any way, so this came as something of a surprise.

I really did some shit with Arcane Affinity, and call it a part of Clarity rework, since Clarity effect is the one and only reason of having Arcane Affinity, and mostly the reason of having ONLY Arcane affinity. Which also means that this can be disabled in the config and in theory that should help with crash too.

> It is worth noting for the sake of being thorough that I am currently using [this fork](https://github.com/RubilaxXxx/ArsMagica2-5/releases) of AM2.5 (using v1.6.9-RE)for the config option for disabling burnout(as a means if avoiding [this issue](https://github.com/TCLProject/ArsMagica2-5/issues/52).
> 
> I know it is technically an unsupported version of AM2.5 for AM2PG, but I'm hoping if nothing else that you can at least tell me if this is a compatibility issue between the fork of AM2.5 and AM2PG or if it is in fact an issue with AM2PG that could happen regardless of what version of AM2.5 is used. I can set aside a time when no one is playing to specifically test a supported version of AM2.5 if you believe it to be needed.

By the way, these versions are not supported not because they shouldn’t work, but because my lazy ass didn’t test them, and I have no guarantee that they work as they should.
As you can see, I didn't test this mod well, especially on the server.
# About rework
That part of rework does this:
when your Arcane Affinity is 50% and above, every 10 seconds (if ticks not skipped) it will randomly give you one of potions from some list, the level of which is calculated randomly, based on current Arcane Affinity %.

**Comment by AirBurn0** [comment]
Well, feel free to test [new pre-release](https://github.com/AirBurn0/am2playground/releases/tag/v0.11.4-beta5) with every version of AM2 you like.

**Comment by Not-February** [comment]
I've tested the pre-release you linked on a server and can confirm that 50% arcane affinity works just fine with it. I'm getting the buffs and not getting kicked, so that's a good sign!

Sadly I won't be able to actually keep it, as we are in fact using a currently-released version of Deathly Hallows. A cooler (and functional) arcane affinity is something to look forward to in the future, I guess!

**Comment by AirBurn0** [comment]
> Sadly I won't be able to actually keep it, as we are in fact using a currently-released version of Deathly Hallows.
I'll see if I can produce new Deathly Hallows version faster.

Considered solved.
